### PR TITLE
Improve performance of block dragging.  This is a backport of the blo…

### PR DESCRIPTION
**Pull request by picklesrus**
…ck drag surface from scratch-blocks.  At the beginning of a block drag, blocks get moved to a drag surface which then translates using translate3d to avoid repainting the entire svg on every mouse move.  At the end of the drag, the blocks are dropped back in the svg in their new position.

### CODE REVIEW

File: core/block_svg.js [hunk 1]
```diff
@@ -342,15 +364,75 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
...
 
 /**
+ * Transforms a block by setting the translation on the transform attribute
+ * of the block's svg.
+ * @param {number} x The x coordinate of the translation.
+ * @param {number} y The y coordinate of the translation.
+ */
+Blockly.BlockSvg.prototype.translate = function(x, y) {
+  this.getSvgRoot().setAttribute('transform',
+    'translate(' + x + ',' + y + ')');
```
Comment (NeilFraser) [inline core/block_svg.js]: More indent.

File: core/block_svg.js [hunk 2]
```diff
@@ -342,15 +364,75 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
...
+};
+
+/**
+ * Clear the block of style="..." and transform="..." attributes.
+ * Used when the block is switching from 3d to 2d transform or vice versa.
+ * @private
+ */
+Blockly.BlockSvg.prototype.clearTransformAttributes_ = function() {
+   if (this.getSvgRoot().hasAttribute('transform')) {
+      this.getSvgRoot().removeAttribute('transform');
```
Comment (NeilFraser) [inline core/block_svg.js]: Less indent.  Here and next line.

File: core/block_svg.js [hunk 3]
```diff
@@ -342,15 +364,75 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
...
+   }
+  if (this.getSvgRoot().hasAttribute('style')) {
+    this.getSvgRoot().removeAttribute('style');
```
Comment (NeilFraser) [inline core/block_svg.js]: Attempting to remove an attribute that is not on the element doesn't raise an exception and may be faster than checking.  Here and above.

File: core/blockly.js [hunk 4]
```diff
@@ -125,6 +125,13 @@ Blockly.dragMode_ = Blockly.DRAG_NONE;
 Blockly.flyoutButtonCallbacks_ = {};
 
 /**
+ * Cached value for whether 3D is supported
+ * @type {boolean}
```
Comment (NeilFraser) [inline core/blockly.js]: !boolean, since it starts null.
Better yet, set it false here, then overwrite?

File: core/css.js [hunk 5]
```diff
@@ -147,6 +147,7 @@ Blockly.Css.CONTENT = [
   '.injectionDiv {',
     'height: 100%;',
     'position: relative;',
+    'overflow:hidden', /* So blocks in drag surface disappear at edges */
```
Comment (NeilFraser) [inline core/css.js]: Space after colon, and also needs semicolon.

File: core/css.js [hunk 6]
```diff
@@ -447,6 +463,7 @@ Blockly.Css.CONTENT = [
     'overflow-x: visible;',
     'overflow-y: auto;',
     'position: absolute;',
+    'z-index:70;',
```
Comment (NeilFraser) [inline core/css.js]: Space needed.

File: core/inject.js [hunk 7]
```diff
@@ -27,6 +27,7 @@
 goog.provide('Blockly.inject');
 
 goog.require('Blockly.Css');
+goog.require('Blockly.BlockDragSurfaceSvg');
```
Comment (NeilFraser) [inline core/inject.js]: Should be alphabetical.

File: core/utils.js [hunk 8]
```diff
@@ -719,3 +739,43 @@ Blockly.utils.wrapToText_ = function(words, wordBreaks) {
...
+ * and attempting to set the property.
+ * @return {boolean} true if 3D transforms are supported
+ */
+Blockly.is3dSupported = function() {
+  if (Blockly.cache3dSupported_ !== null) {
+    return Blockly.cache3dSupported_;
+  }
+  // CC-BY-SA Lorenzo Polidori
+  // https://stackoverflow.com/questions/5661671/detecting-transform-translate3d-support
+  if (!window.getComputedStyle) {
```
Comment (NeilFraser) [inline core/utils.js]: Use goog.global instead of window.  NodeJS and non-browser environments don't have window.

File: core/utils.js [hunk 9]
```diff
@@ -719,3 +739,43 @@ Blockly.utils.wrapToText_ = function(words, wordBreaks) {
...
+
+  for (var t in transforms) {
+    if (el.style[t] !== undefined) {
+      el.style[t] = 'translate3d(1px,1px,1px)';
+      has3d = window.getComputedStyle(el).getPropertyValue(transforms[t]);
+    }
+  }
+
+  document.body.removeChild(el);
+  Blockly.cache3dSupported_ = (has3d !== undefined && has3d.length > 0 && has3d !== "none");
```
Comment (NeilFraser) [inline core/utils.js]: Suggestion:
Blockly.cache3dSupported_ = !!(has3d && has3d !== 'none');

File: core/utils.js [hunk 10]
```diff
@@ -719,3 +739,43 @@ Blockly.utils.wrapToText_ = function(words, wordBreaks) {
...
+    transforms = {
+      'webkitTransform': '-webkit-transform',
+      'OTransform': '-o-transform',
+      'msTransform': '-ms-transform',
+      'MozTransform': '-moz-transform',
+      'transform': 'transform'
+    };
+
+  // Add it to the body to get the computed style.
+  document.body.insertBefore(el, null);
```
Comment (NeilFraser) [inline core/utils.js]: Here's an underused function you might like:
https://developer.mozilla.org/en-US/docs/Web/API/Node/replaceChild

File: core/workspace_svg.js [hunk 11]
```diff
@@ -48,18 +48,24 @@ goog.require('goog.userAgent');
...
       options.getMetrics || Blockly.WorkspaceSvg.getTopLevelWorkspaceMetrics_;
   this.setMetrics =
       options.setMetrics || Blockly.WorkspaceSvg.setTopLevelWorkspaceMetrics_;
 
   Blockly.ConnectionDB.init(this);
-
+ 
+ if (opt_blockDragSurface) {
+    this.blockDragSurface_ = opt_blockDragSurface;
+
```
Comment (NeilFraser) [inline core/workspace_svg.js]: Unneeded blank line.

File: core/blockly.js (hunk 4 above)
Comment (picklesrus) [inline core/blockly.js]: I need to be able to tell the difference between has been checked and hasn't because is3dSupported check is expensive. It is cacheable though, thus the initialization to null as an indicator that we haven't run the check yet.

File: core/utils.js (hunk 10 above)
Comment (picklesrus) [inline core/utils.js]: Oh neat. Thanks. Not sure it is applicable here since this isn't really replacing anything.  It is inserting the note into the dom to check whether it can perform certain 3d stuff, then removing it.

**Comment by picklesrus** [comment]
Thanks. And sorry I forgot to run lint. Fixed a few more things lint found as well. Think I got 'em all.

Also edge is now fixed, but this approach is slower in Edge so we may decide to disable it.

File: core/utils.js (hunk 10 above)
Comment (NeilFraser) [inline core/utils.js]: Ah, my mistake.  Didn't read the code carefully enough.

**Comment by picklesrus** [comment]
Did a little more performance testing in Edge. It is fairly bad both for the existing code and with this change.  With one or two blocks, this definitely makes it worse. With 100+, they're both terrible. And with 10-20, they're pretty similarly bad.

All testing done on Surface 3 so a pretty low end device. For consistency it would be nice to have this change on everywhere.  My vote is to check this in with edge every once in a while and see if the performance of diverges.

File: core/block_svg.js [hunk 12]
```diff
@@ -342,15 +364,71 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
...
+  this.workspace.blockDragSurface_.clearAndHide(this.workspace.getCanvas());
+};
+
+/**
+ * Clear the block of style="..." and transform="..." attributes.
+ * Used when the block is switching from 3d to 2d transform or vice versa.
+ * @private
+ */
+Blockly.BlockSvg.prototype.clearTransformAttributes_ = function() {
+  this.getSvgRoot().removeAttribute('transform');
```
Comment (tmickel) [inline core/block_svg.js]: I forget, did we decide that this is unnecessary, or not?

When we move blocks to the drag surface, I think they should maintain their `transform` attribute, not switch to `style` (as was a previous implementation). If that's the case, then I think we can drop this function and the calls to it.

File: core/css.js [hunk 13]
```diff
@@ -156,6 +157,21 @@ Blockly.Css.CONTENT = [
...
+    'top: 0;',
+    'left: 0;',
+    'right: 0;',
+    'bottom: 0;',
+    'overflow: visible !important;',
+    'z-index: 50;', /* Display below toolbox, but everything else. */
+    '-webkit-backface-visibility: hidden;',
+    'backface-visibility: hidden;',
+    '-webkit-perspective: 1000;',
+    'perspective: 1000;',
```
Comment (tmickel) [inline core/css.js]: I would double check, but I'm not confident whether these CSS properties (backface visibility, perspective) are necessary or just left-over from experiments.

File: core/block_svg.js (hunk 12 above)
Comment (picklesrus) [inline core/block_svg.js]: The style one isn't necessary. The transform once is.  This is the transform is the one on the block itself, which is part of what getRelativeToSurfaceXY returns.  If we don't clear it, we end up adding it in twice. I guess we could change getRelativeToSurfaceXY, except that it is used in a bunch more places.  Instead I guess we could add another helper that doesn't take into account the transform instead?  But, clearing it here instead seems simpler to me.

File: core/css.js (hunk 13 above)
Comment (picklesrus) [inline core/css.js]: Ah, good call. Removed. Don't see a difference.

File: core/block_drag_surface.js [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * Translate and scale the entire drag surface group to keep in sync with the workspace.
+ * @param {number} x X translation
+ * @param {number} y Y translation
+ * @param {number} scale Scale of the group.
+ */
+Blockly.BlockDragSurfaceSvg.prototype.translateAndScaleGroup = function(x, y, scale) {
+  this.scale_ = scale;
+  // Force values to have two decimal points.
```
Comment (picklesrus) [inline core/block_drag_surface.js]: Fixed

File: core/utils.js [hunk 15]
```diff
@@ -266,6 +274,18 @@ Blockly.getRelativeXY_ = function(element) {
       xy.y += parseFloat(r[3]);
     }
   }
+
+    // Third, check for style="transform: translate3d(...)".
+  var style = element.getAttribute('style');
```
Comment (picklesrus) [inline core/utils.js]: Fixed (I think).

File: core/block_drag_surface.js [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Functions for injecting Blockly into a web page.
```
Comment (rachel-fenichel) [inline core/block_drag_surface.js]: @fileoverview is wrong.

File: core/block_drag_surface.js [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+'use strict';
+
+goog.provide('Blockly.BlockDragSurfaceSvg');
+goog.require('Blockly.utils');
+goog.require('goog.asserts');
+goog.require('goog.math.Coordinate');
+
+
+/**
+ * Class for a Drag Surface SVG.
```
Comment (rachel-fenichel) [inline core/block_drag_surface.js]: Class for a drag surface for the currently dragged block.  This is a separate SVG that contains only the currently moving block, or nothing.

File: core/block_drag_surface.js [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+ * @param {!Element} container Containing element.
+ * @constructor
+ */
+Blockly.BlockDragSurfaceSvg = function(container) {
+  this.container_ = container;
```
Comment (rachel-fenichel) [inline core/block_drag_surface.js]: @type tag here, and @private.

File: core/block_drag_surface.js [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+
+/**
+ * The SVG drag surface. Set once by Blockly.BlockDragSurfaceSvg.createDom.
+ * @type {Element}
+ * @private
+ */
+Blockly.BlockDragSurfaceSvg.prototype.SVG_ = null;
+
+/**
+ * SVG group inside the drag surface. This is where blocks are moved to.
```
Comment (rachel-fenichel) [inline core/block_drag_surface.js]: This is where blocks live while they are being dragged.

File: core/block_drag_surface.js [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+    'xmlns:xlink': 'http://www.w3.org/1999/xlink',
+    'version': '1.1',
+    'class': 'blocklyBlockDragSurface'
+  }, this.container_);
+  this.dragGroup_ = Blockly.createSvgElement('g', {}, this.SVG_);
+};
+
+/**
+ * Set the SVG blocks on the drag surface's group and show the surface.
+ * Only one block should be on the drag surface at a time.
```
Comment (rachel-fenichel) [inline core/block_drag_surface.js]: Only one block group should be on the drag surface at a time.

File: core/block_drag_surface.js [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+  this.SVG_.style.display = 'block';
+};
+
+/**
+ * Translate and scale the entire drag surface group to keep in sync with the workspace.
+ * @param {number} x X translation
+ * @param {number} y Y translation
+ * @param {number} scale Scale of the group.
+ */
+Blockly.BlockDragSurfaceSvg.prototype.translateAndScaleGroup = function(x, y, scale) {
```
Comment (rachel-fenichel) [inline core/block_drag_surface.js]: If this should only be called at the beginning or end of a drag, include that in the comment.

File: core/block_svg.js [hunk 22]
```diff
@@ -342,15 +364,70 @@ Blockly.BlockSvg.prototype.moveBy = function(dx, dy) {
...
+  var xy = this.getRelativeToSurfaceXY();
+  this.clearTransformAttributes_();
+  this.workspace.blockDragSurface_.translateSurface(xy.x, xy.y);
+  // Execute the move on the top-level SVG component
+  this.workspace.blockDragSurface_.setBlocksAndShow(this.getSvgRoot());
+};
+
+/**
+ * Move this block back to the workspace block canvas.
+ * Generally should be called at the same time as setDragging_(false).
```
Comment (rachel-fenichel) [inline core/block_svg.js]: Does nothing if useDragSurface_ is false.

File: core/css.js [hunk 23]
```diff
@@ -156,6 +157,17 @@ Blockly.Css.CONTENT = [
...
 
+  '.blocklyBlockDragSurface {',
+    'display: none;',
+    'position: absolute;',
+    'top: 0;',
+    'left: 0;',
+    'right: 0;',
+    'bottom: 0;',
+    'overflow: visible !important;',
+    'z-index: 50;', /* Display below toolbox, but everything else. */
```
Comment (rachel-fenichel) [inline core/css.js]: Display below toolbox, but above everything else.

File: core/utils.js [hunk 24]
```diff
@@ -238,6 +238,14 @@ Blockly.isTargetInput_ = function(e) {
          e.target.type == 'tel' || e.target.type == 'url' ||
          e.target.isContentEditable;
 };
+/**
```
Comment (rachel-fenichel) [inline core/utils.js]: Newline above this.

File: core/workspace_svg.js [hunk 25]
```diff
@@ -509,6 +522,9 @@ Blockly.WorkspaceSvg.prototype.translate = function(x, y) {
       'scale(' + this.scale + ')';
   this.svgBlockCanvas_.setAttribute('transform', translation);
   this.svgBubbleCanvas_.setAttribute('transform', translation);
+  if (this.blockDragSurface_) {
+    this.blockDragSurface_.translateAndScaleGroup(x, y, this.scale);
```
Comment (rachel-fenichel) [inline core/workspace_svg.js]: Does this need to be updated every time the workspace translates?

I guess the advantage is that we don't have this slowing down block drag starts, but doesn't it make every workspace movement slower?

Comment (tmickel) [inline core/workspace_svg.js]: I believe it basically makes no difference. You could move it to the block drag start, which would add some trivial amount of time there - but should be insignificant compared to the repaint which is about to happen. OTOH, calling it whenever the workspace translates and scales should also be very fast, since the element is hidden and there should be nothing in the SVG. I originally put it here because it made most sense to me conceptually as the goal is to match the workspace translation/scale.

File: core/block_drag_surface.js (hunk 21 above)
Comment (tmickel) [inline core/block_drag_surface.js]: if the switch below is made, I think it doesn't need to be called at the end of a drag, only the beginning.

File: core/workspace_svg.js (hunk 25 above)
Comment (rachel-fenichel) [inline core/workspace_svg.js]: Sweet--if it's fast because it's hidden, then that's fine.

File: core/block_drag_surface.js (hunk 19 above)
Comment (tmickel) [inline core/block_drag_surface.js]: if the drag surface is enabled :)

**Comment by tmickel** [comment]
LGTM!!

**Comment by picklesrus** [comment]
Sweet. I'll merge once the develop->master merge (PR #747) lands.

**Comment by rachel-fenichel** [comment]
https://i.imgur.com/eMqtr39.gif
